lib/streamDAO.js

```diff
@@ -254,24 +232,57 @@ module.exports = function(mongoClient){
 
       datum = ensureId(datum);
 
-      mongoClient.withCollection('deviceData', cb, function(coll, cb){
-        coll.insertOne(datum, function(err){
-          if (err != null) {
-            if (err.code === 11000) {
-              return cb({ statusCode: 400, errorCode: 'duplicate', message: 'received a duplicate event'});
-            }
-            return cb(err);
+      const errHandler = (err) => {
+        if (err != null) {
+          if (err.code === 11000) {
+            return cb({ statusCode: 400, errorCode: 'duplicate', message: 'received a duplicate event'});
           }
-          return cb(null, datum);
+          return cb(err);
+        }
+        cb(null, datum);
+      };
+
+      if (datum.type == 'upload') {
+        const insertInTx = async (session) => {
+          // We need to insert a dataSet (datum's with type == "upload") into
+          // both the deviceData and deviceDataSets collection because as the
+          // migration from deviceData to deviceDataSets happens, some not
+          // yet updated clients might still expect dataSet to be in the old
+          // deviceData collection.
+          await mongoClient.collection('deviceData').insertOne(datum, {session});
+          await mongoClient.collection('deviceDataSets').insertOne(datum, {session});
+        };
+        mongoClient.transact(insertInTx, errHandler);
+      }
+      else {
+        mongoClient.withCollection(getCollectionName(datum), cb, function(coll, cb){
+          coll.insertOne(datum, errHandler);
         });
-      });
+      }
     },
     updateDatum: function(datum, cb) {
       pre.hasProperty(datum, 'id');
       pre.hasProperty(datum, '_userId');
       pre.hasProperty(datum, '_groupId');
       var filteredDatum = filterDatumForMongo(datum);
-      updateDatumInternal(filteredDatum, 0, cb);
+
+      let firstResult = null;
+      const collections = getCollections(datum);
+      // In order to handle concurrent updates / avoid lost updates we will
+      // always run this in a transaction to mimick the behaviour of the
+      // previous updateDatumInternal code.
+      const updateInTx = async (session) => {
+        for (let collectionName of collections) {
+          const collection = mongoClient.collection(collectionName);
+          const result = await updateDatumInternal(session, collection, filteredDatum);
```

Comment:
The previous version of `updateDatumInternal` used to retry the operation on collision. Now this logic is removed, but I'm not sure what the implications are. Any thoughts?

Reply:
I believe the automatic retry of the callback captures this logic in a simpler fashion. I will try to test this specific code locally.